monkey/infection_monkey/monkey.py

```diff
@@ -411,11 +413,20 @@ def _stop_relay(self):
         if self._relay and self._relay.is_alive():
             self._relay.stop()
 
-            while self._relay.is_alive() and not self._control_channel.should_agent_stop():
-                self._relay.join(timeout=5)
+            while self._relay.is_alive():
+                self._wait_for_clients_shutdown()
 
-            if self._control_channel.should_agent_stop():
-                self._relay.join(timeout=60)
+            self._wait_for_clients_killed()
+
+    @retry(times=5, exceptions=[IslandCommunicationError])
+    def _wait_for_clients_shutdown(self, iteration_timeout=5):
+        if not self._control_channel.should_agent_stop():
+            self._relay.join(timeout=iteration_timeout)
```

Comment:
This is a fundamental logic change that could cause issues.

The `not self._control_channel.should_agent_stop()` needs to be part of the loop condition, otherwise this could lead to agents hanging and not shutting down properly when the terminate signal is sent.

If you examine the old logic, once the terminate signal is sent, the loop exits and the next check will wait a maximum of 60 seconds for the relay to stop before shutting down. This change causes the loop to run infinitely until the relay shuts down. In the old logic, the relay will be forced to shut down when the main thread exits after 60 seconds, which is desirable.